src/tool/abi/common.h

```diff
@@ -0,0 +1,342 @@
+        {
+            return cmp < 0;
+        }
+        else
+        {
+            // Same namespace
+            return name < other.name;
+        }
+    }
+
+    // So that this type can be used in contexts that call 'TypeName' and 'TypeNamespace'
+    auto TypeNamespace() { return this->ns; }
+    auto TypeName() { return this->name; }
+};
+
+struct typename_compare
+{
+    template <typename LhsT, typename RhsT>
+    bool operator()(LhsT const& lhs, RhsT const& rhs) const noexcept
+    {
+        auto cmp = lhs.TypeNamespace().compare(rhs.TypeNamespace());
+        if (cmp == 0)
+        {
+            cmp = lhs.TypeName().compare(rhs.TypeName());
+        }
+
+        return cmp < 0;
+    }
+};
+
+enum class ns_prefix
+{
+    always,
+    never,
+    optional,
+};
+
+struct abi_configuration
+{
+    bool verbose = false;
+    ns_prefix ns_prefix_state = ns_prefix::always;
+    bool enum_class = false;
+    bool lowercase_include_guard = false;
+
+    std::string output_directory;
+
+    std::pair<bool, type_name> map_type(std::string_view typeNamespace, std::string_view typeName) const noexcept
+    {
+        using namespace std::literals;
+
+        bool isMapped = false;
+        if (typeNamespace == foundation_namespace)
+        {
+            if (typeName == "EventRegistrationToken"sv)
+            {
+                // Windows.Foundation.EventRegistrationToken -> EventRegistrationToken
+                isMapped = true;
+                typeNamespace = {};
+            }
+            else if (typeName == "IAsyncInfo"sv)
+            {
+                // Windows.Foundation.IAsyncInfo -> IAsyncInfo
+                isMapped = true;
+                typeNamespace = {};
+            }
+            else if (typeName == "AsyncStatus"sv)
+            {
+                // Windows.Foundation.AsyncStatus -> AsyncStatus
+                isMapped = true;
+                typeNamespace = {};
+            }
+            else if (typeName == "HResult"sv)
+            {
+                // Windows.Foundation.HResult -> HRESULT
+                isMapped = true;
+                typeNamespace = {};
+                typeName = "HRESULT"sv;
+            }
+        }
+
+        return { isMapped, { typeNamespace, typeName } };
+    }
+};
+
+namespace xlang
+{
+    template <typename... T> visit_overload(T...) -> visit_overload<T...>;
+}
+
+template <typename Visitor>
+inline auto visit(xlang::meta::reader::coded_index<xlang::meta::reader::TypeDefOrRef> const& type, Visitor&& visitor)
+{
+    using namespace xlang::meta::reader;
+    switch (type.type())
+    {
+    case TypeDefOrRef::TypeDef: return visitor(type.TypeDef()); break;
+    case TypeDefOrRef::TypeRef: return visitor(type.TypeRef()); break;
+    case TypeDefOrRef::TypeSpec: return visitor(type.TypeSpec().Signature().GenericTypeInst()); break;
+    }
+}
+
+inline constexpr std::pair<std::string_view, std::string_view> decompose_type(std::string_view typeName) noexcept

```

Comment:
constexpr functions are implicitly inline.

Reply:
heh, habit mostly. Shouldn't be harmful